lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -0,0 +1,705 @@
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "DxilValidationUtils.h"
+
+#include <memory>
+
+using std::unique_ptr;
+using std::unordered_set;
+using std::vector;
+
+namespace {
+
+// Utility class for setting and restoring the diagnostic context so we may
+// capture errors/warnings
+struct DiagRestore {
+  LLVMContext *Ctx = nullptr;
+  void *OrigDiagContext;
+  LLVMContext::DiagnosticHandlerTy OrigHandler;
+
+  DiagRestore(llvm::LLVMContext &InputCtx, void *DiagContext) : Ctx(&InputCtx) {
+    init(DiagContext);
+  }
+  DiagRestore(Module *M, void *DiagContext) {
+    if (!M)
+      return;
+    Ctx = &M->getContext();
+    init(DiagContext);
+  }
+  ~DiagRestore() {
+    if (!Ctx)
+      return;
+    Ctx->setDiagnosticHandler(OrigHandler, OrigDiagContext);
+  }
+
+private:
+  void init(void *DiagContext) {
+    OrigHandler = Ctx->getDiagnosticHandler();
+    OrigDiagContext = Ctx->getDiagnosticContext();
+    Ctx->setDiagnosticHandler(
+        hlsl::PrintDiagnosticContext::PrintDiagnosticHandler, DiagContext);
+  }
+};
+
+static void emitDxilDiag(LLVMContext &Ctx, const char *str) {
+  hlsl::dxilutil::EmitErrorOnContext(Ctx, str);
+}
+} // namespace
+
+namespace hlsl {
+
+// DXIL Container Verification Functions
+
+static void VerifyBlobPartMatches(ValidationContext &ValCtx, LPCSTR pName,
+                                  DxilPartWriter *pWriter, const void *pData,
+                                  uint32_t Size) {
+  if (!pData && pWriter->size()) {
+    // No blob part, but writer says non-zero size is expected.
+    ValCtx.EmitFormatError(ValidationRule::ContainerPartMissing, {pName});
+    return;
+  }
+
+  // Compare sizes
+  if (pWriter->size() != Size) {
+    ValCtx.EmitFormatError(ValidationRule::ContainerPartMatches, {pName});
+    return;
+  }
+
+  if (Size == 0) {
+    return;
+  }
+
+  CComPtr<AbstractMemoryStream> pOutputStream;
+  IFT(CreateMemoryStream(DxcGetThreadMallocNoRef(), &pOutputStream));
+  pOutputStream->Reserve(Size);
+
+  pWriter->write(pOutputStream);
+  DXASSERT(pOutputStream->GetPtrSize() == Size,
+           "otherwise, DxilPartWriter misreported size");
+
+  if (memcmp(pData, pOutputStream->GetPtr(), Size)) {
+    ValCtx.EmitFormatError(ValidationRule::ContainerPartMatches, {pName});
+    return;
+  }
+
+  return;
+}
+
+static void VerifySignatureMatches(ValidationContext &ValCtx,
+                                   DXIL::SignatureKind SigKind,
+                                   const void *pSigData, uint32_t SigSize) {
+  // Generate corresponding signature from module and memcmp
+
+  const char *pName = nullptr;
+  switch (SigKind) {
+  case hlsl::DXIL::SignatureKind::Input:
+    pName = "Program Input Signature";
+    break;
+  case hlsl::DXIL::SignatureKind::Output:
+    pName = "Program Output Signature";
+    break;
+  case hlsl::DXIL::SignatureKind::PatchConstOrPrim:
+    if (ValCtx.DxilMod.GetShaderModel()->GetKind() == DXIL::ShaderKind::Mesh)
```

Comment:
Should we check for node shaders here? I think mesh nodes can have patch constant functions right?

Reply:
We don't have a Signature part for node shaders.